### DISCUSSION TITLE
Iterating over tagged_iterators in constructor will segfault php

### DISCUSSION THREAD

**Discussion by jaytaph** [symfony/symfony#48788] [Q&A] [answered]
### Symfony version(s) affected

6.2

### Description

Suppose we have a service called `QueueService`, this service collects `WorkerInterface` classes through the standard tagged_iterators method. Each of these workerinterface classes will have the `QueueService` as an argument.

At this point, when we iterate over the tagged iterators (`RewindableGenerator`), the elements in this generator will get resolved. The problem here is that this means the QueueService will get resolved too, as it is not marked as being resolved yet (since we are still inside the __construct of the queueservice).  This will create an endless loop as we call the __construct again, which in turn will walk over the workerinterfaces, which will call the Queueservice constructor again etc... until we probably run out of stack and php will segfault.

Since the symfony docs uses "iterator_to_array()" in the documentation to convert tagged_iteartors from an iterable to an array, this is bound to happen more often.

### How to reproduce

https://github.com/jaytaph/symfony-iterator-bug

### Possible Solution

If we mark the fact that we are currently resolving a service, we can detect these kind of loops.

**Comment by nicolas-grekas** [comment, marked as answer]
Such segfault in PHP are very likely due to stack overflows. Aka the function recursively calls itself *ad nauseam*.
While the error is not really helpful (PHP 8.3 might improve this BTW), the issue is definitely on your side.
You might want to [exclude the current class from the iterator](https://github.com/symfony/symfony/pull/44774) (that's the most likely cause to me.)

Alternatively, you might want to *not* iterate in the constructor.

**Reply by chalasr** [reply]
#48685 should help by preventing that recursion by default
